connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/StreamingQueryListenerHelper.scala

```diff
@@ -76,16 +78,21 @@ class PythonStreamingQueryListener(listener: SimplePythonFunction, sessionHolder
   }
 
   private def handlePythonWorkerError(functionName: String): Unit = {
-    dataIn.readInt() match {
-      case ret if ret == 0 =>
-        logInfo(s"Streaming query listener function $functionName completed (ret: $ret)")
-      case SpecialLengths.PYTHON_EXCEPTION_THROWN =>
-        val exLength = dataIn.readInt()
-        val obj = new Array[Byte](exLength)
-        dataIn.readFully(obj)
-        val msg = new String(obj, StandardCharsets.UTF_8)
-        throw new IllegalStateException(s"Found error inside Streaming query listener Python " +
-          s"process for function $functionName: $msg")
+    try {
+      dataIn.readInt() match {
+        case ret if ret == 0 =>
+          logInfo(s"Streaming query listener function $functionName completed (ret: $ret)")
+        case SpecialLengths.PYTHON_EXCEPTION_THROWN =>
+          val exLength = dataIn.readInt()
+          val obj = new Array[Byte](exLength)
+          dataIn.readFully(obj)
+          val msg = new String(obj, StandardCharsets.UTF_8)
+          throw new PythonException(s"Found error inside Streaming query listener Python " +
+            s"process for function $functionName: $msg", null)
+      }
+    } catch {
+      case eof: EOFException =>
```

Comment:
@HyukjinKwon @taku-k Guys, do you know if the EOFException would happen frequently in existing pyspark tests? I was wondering if this is possible:
Say in common PythonRunner, EOF occurs, then the task would likely be failing but spark nicely handles the retry.

But here if the socket has some error, we don't have any retry logic, and the stream query would just fail with our issue.

Given that streamingPythonRunners are long-running processes, I was wondering if you also think having a retry-logic would be beneficial as a followup.

cc @rangadi

Reply:
If we want to add retry logic, adding it here would not solve the problem. If the EOF is due to an IOException during `write_` functions, we might need to add retries inside the workers. But I'm not sure how high would the possibility be for that scenario to happen, since if IOException happens while writing, it's highly possible that the socket is closed, and retry would no help in that case.

Reply:
Yeah, we can't handle the retries in the PR. we need to implement that feature properly.
@bogao007 can catch any `NonFatal(ex)` here.

Reply:
Not sure if catching any `NonFatal(ex)` is a good solution here since if errors occurs, it's highly possible that something went wrong in the python worker, and better to fail the query in this case. I'll leave it as what it is for now and also added a TODO in `StreamingForeachBatchHelper` to improve handling this scenario later.

Reply:
Absolutely. It should fail the query. I thought this code was wrapping it in a SparkException and throwing it.